packages/react/src/components/form/ui-state-machine.ts

```diff
@@ -265,6 +297,111 @@ class PasswordUIStateMachine extends UIStateMachine<PasswordStatus> {
   }
 }
 
+class TOTPUIStateMachine extends UIStateMachine<TOTPStatus> {
+  private recoveryCodes: string[] | null;
+  private isRegisterFlow: boolean;
+
+  constructor(opts: UIStateMachineOpts) {
+    const totpKeyGeneratedHandler = ({
+      qrCode,
+      recoveryCodes,
+      uri,
+    }: TotpKeyGenerated) => {
+      this.recoveryCodes = recoveryCodes;
+      this.isRegisterFlow = true;
+
+      this.send({
+        type: "sid_ui.totpKeyGenerated",
+        qrCode,
+        uri,
+      });
+    };
+    const totpCodeRequestedHandker = () => {
+      this.send({ type: "sid_ui.totpCodeRequested" });
+    };
+
+    super(opts, {
+      status: "initial",
+      entry: () => {
+        this.sid.subscribe("totpKeyGenerated", totpKeyGeneratedHandler);
+        this.sid.subscribe("totpCodeRequested", totpCodeRequestedHandker);
+
+        this.logInFn(this.context.config, this.context.options)
+          .then((user) => {
+            if (user) {
+              if (this.isRegisterFlow) {
+                this.send({ type: "sid_ui.userAuthenticated", user });
+              } else {
+                this.send({ type: "sid_login.success", user });
+              }
+            } else {
```

Comment:
An early return in case the user is not returned here would help you get rid of some indentation